File: System/Database/Allors.Database/Data/Procedure.cs
// <copyright file="Sort.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Database.Data
{
    using System.Collections.Generic;

    public class Procedure : IVisitable
    {
        public Procedure(string name) => this.Name = name;

        public string Name { get; }

        public IDictionary<string, IObject[]> CollectionByName { get; set; }

        public IDictionary<string, IObject> ObjectByName { get; set; }

        public IDictionary<string, object> ValueByName { get; set; }

        public IDictionary<IObject, long> VersionByObject { get; set; }

        public void Accept(IVisitor visitor) => visitor.VisitProcedure(this);
    }
}

File: typescript/libs/protocol/json/system/src/lib/data/Procedure.ts
import { UnitTypes } from "@allors/workspace/system";

export interface Procedure {
  /** Name */
  n: string;

  /** Collections */
  c: { [name: string]: number[] };

  /** Objects */
  o: { [name: string]: number };

  /** Values */
  v: { [name: string]: UnitTypes };

  /** Pool
   *  [][id,version]
   */
  p: number[][];
}
